crates/stages/src/stages/merkle.rs

```diff
@@ -33,15 +38,47 @@ impl<DB: Database> Stage<DB> for MerkleStage {
     /// Execute the stage.
     async fn execute(
         &mut self,
-        _tx: &mut Transaction<'_, DB>,
+        tx: &mut Transaction<'_, DB>,
         input: ExecInput,
     ) -> Result<ExecOutput, StageError> {
         if !self.is_execute {
             info!(target: "sync::stages::merkle::unwind", "Stage is always skipped");
             return Ok(ExecOutput { stage_progress: input.previous_stage_progress(), done: true })
         }
 
-        // Iterate over changeset (similar to Hashing stages) and take new values
+        let stage_progress = input.stage_progress.unwrap_or_default();
+        let previous_stage_progress = input.previous_stage_progress();
+
+        let from_transition = tx.get_block_transition(stage_progress)?;
+        let to_transition = tx.get_block_transition(previous_stage_progress)?;
+
+        let previous_numhash = tx.get_block_numhash(previous_stage_progress)?;
+
+        let block_root = tx
+            .get::<tables::Headers>(previous_numhash)?
+            .ok_or_else(|| {
+                StageError::DatabaseIntegrity(DatabaseIntegrityError::Header {
+                    number: previous_stage_progress,
+                    hash: previous_numhash.hash(),
+                })
+            })?
+            .state_root;
+
+        let trie_root = if to_transition - from_transition > self.clean_threshold {
+            // if there are more blocks then threshold it is faster to rebuild the trie
+            let mut loader = DBTrieLoader {};
+            loader.calculate_root(tx).map_err(|e| StageError::Fatal(Box::new(e)))?
```

Comment:
Full state hashing is not just about calculating root but calculating all intermediate hashes and saving them to table as they are needed for partial update of the trie in the future.

Reply:
Yes. This would be done inside `calculate_root` (although it's not implemented yet).